File: Tests/Strata.Foundation.CoreTest/test/typescript/Mapper/PredicatedMapperTest.ts
import "jest"
import {Optional}  from "strata.foundation.core/Utility";
import {PredicatedMapper} from "strata.foundation.core/Mapper";

describe(
    "PredicatedMapperTest",
    () =>
    {
        it(
            "testGet",
            () =>
            {
                const subject: PredicatedMapper<string,string> =
                    PredicatedMapper
                        .of<string,string>("input X")
                        .addMapping(
                            (input) => input == "input A",
                            (input) => "mapped A")
                        .addMapping(
                            (input) => input == "input B",
                            (input) => "mapped B")
                        .addMapping(
                            (input) => input == "input X",
                            (input) => "mapped X")
                        .addMapping(
                            (input) => input == "input C",
                            (input) => "mapped C");

                const expected: string = "mapped X";
                const actual: Optional<string> = subject.get();

                expect(actual.isPresent()).toBeTruthy();
                expect(actual.get()).toBe(expected);

            });
        it(
            "testGetWhenEmpty",
            () =>
            {
                const subject: PredicatedMapper<string,string> =
                    PredicatedMapper
                        .of<string,string>("input X")
                        .addMapping(
                            (input) => input == "input A",
                            (input) => "mapped A")
                        .addMapping(
                            (input) => input == "input B",
                            (input) => "mapped B")
                        .addMapping(
                            (input) => input == "input C",
                            (input) => "mapped C");

                const expected: string = "mapped X";
                const actual: Optional<string> = subject.get();

                expect(actual.isPresent()).toBeFalsy();
            });
        it(
            "testMapInputFromConstructor",
            () =>
            {
                const subject: PredicatedMapper<string,string> =
                    PredicatedMapper
                        .of<string,string>("input X")
                        .addMapping(
                            (input) => input == "input A",
                            () => "mapped A")
                        .addMapping(
                            (input) => input == "input B",
                            () => "mapped B")
                        .addMapping(
                            (input) => input == "input X",
                            () => "mapped X")
                        .addMapping(
                            (input) => input == "input C",
                            () => "mapped C");

                const expected: string = "mapped X";
                const actual: Optional<string> = subject.map();

                expect(actual.isPresent()).toBeTruthy();
                expect(actual.get()).toBe(expected);

            });
        it(
            "testMapWhenEmptyInputFromConstructor",
            () =>
            {
                const subject: PredicatedMapper<string,string> =
                    PredicatedMapper
                        .of<string,string>("input X")
                        .addMapping(
                            (input) => input == "input A",
                            () => "mapped A")
                        .addMapping(
                            (input) => input == "input B",
                            () => "mapped B")
                        .addMapping(
                            (input) => input == "input C",
                            () => "mapped C");

                const expected: string = "mapped X";
                const actual: Optional<string> = subject.map();

                expect(actual.isPresent()).toBeFalsy();
            });
        it(
            "testMap",
            () =>
            {
                const subject: PredicatedMapper<string,string> =
                    PredicatedMapper
                        .of<string,string>()
                        .addMapping(
                            (input) => input == "input A",
                            () => "mapped A")
                        .addMapping(
                            (input) => input == "input B",
                            () => "mapped B")
                        .addMapping(
                            (input) => input == "input X",
                            () => "mapped X")
                        .addMapping(
                            (input) => input == "input C",
                            () => "mapped C");

                const expected: string = "mapped X";
                const actual: Optional<string> = subject.map("input X");

                expect(actual.isPresent()).toBeTruthy();
                expect(actual.get()).toBe(expected);

            });
        it(
            "testMapWhenEmpty",
            () =>
            {
                const subject: PredicatedMapper<string,string> =
                    PredicatedMapper
                        .of<string,string>()
                        .addMapping(
                            (input) => input == "input A",
                            () => "mapped A")
                        .addMapping(
                            (input) => input == "input B",
                            () => "mapped B")
                        .addMapping(
                            (input) => input == "input C",
                            () => "mapped C");

                const expected: string = "mapped X";
                const actual: Optional<string> = subject.map("input X");

                expect(actual.isPresent()).toBeFalsy();
            });

    });

File: Tests/Strata.Foundation.CoreTest/test/java/strata/foundation/core/mapper/PredicatedMapperTest.java
/// ///////////////////////////////////////////////////////////////////////////
// PredicatedMapperTest.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.mapper;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;

@Tag("CommitStage")
public
class PredicatedMapperTest
{
    @Test
    public void
    testMap()
    {
        PredicatedMapper mapper =
            new PredicatedMapper()
                .beginTypeMap(String.class,String.class)
                .addMapping(
                    input -> input.equals("input A"),
                    input -> "mapped A")
                .addMapping(
                    input -> input.equals("input B"),
                    input -> "mapped B")
                .addMapping(
                    input -> input.equals("input X"),
                    input -> "mapped X")
                .addMapping(
                    input -> input.equals("input C"),
                    input -> "mapped C")
                .beginTypeMap(Long.class,String.class)
                .addMapping(
                    input -> input.equals(7L),
                    input -> "mapped 7")
                .toMapper();

        String expected = "mapped X";
        Optional<String> actual = mapper.map("input X",String.class);

        assertTrue(actual.isPresent());
        assertEquals(expected,actual.get());

        mapper
            .map("inpup X", String.class)
            .ifPresent(actualValue -> assertEquals(expected,actualValue));

        assertTrue(mapper.map(7L, String.class).isPresent());
        mapper
            .map(7L, String.class)
            .ifPresent(actualValue -> assertEquals("mapped 7",actualValue));
    }

    @Test
    public void
    testWhenEmpty()
    {
        PredicatedMapper mapper =
            new PredicatedMapper()
                .beginTypeMap(String.class,String.class)
                .addMapping(
                    input -> input.equals("input A"),
                    input -> "mapped A")
                .addMapping(
                    input -> input.equals("input B"),
                    input -> "mapped B")
                .addMapping(
                    input -> input.equals("input C"),
                    input -> "mapped C")
                .toMapper();

        Optional<String> actual = mapper.map("input X", String.class);

        assertFalse(actual.isPresent());
    }

    @Test
    public void
    testMapWithIncorrectInputType()
    {
        PredicatedMapper mapper =
            new PredicatedMapper()
                .beginTypeMap(String.class,String.class)
                .addMapping(
                    input -> input.equals("input A"),
                    input -> "mapped A")
                .addMapping(
                    input -> input.equals("input B"),
                    input -> "mapped B")
                .addMapping(
                    input -> input.equals("input X"),
                    input -> "mapped X")
                .addMapping(
                    input -> input.equals("input C"),
                    input -> "mapped C")
                .toMapper();

        String expected = "mapped X";
        Optional<String> actual = mapper.map(7L, String.class);

        assertFalse(actual.isPresent());
    }

    @Test
    public void
    testMapWithIncorrectOutputType()
    {
        PredicatedMapper mapper =
            new PredicatedMapper()
                .beginTypeMap(String.class,String.class)
                .addMapping(
                    input -> input.equals("input A"),
                    input -> "mapped A")
                .addMapping(
                    input -> input.equals("input B"),
                    input -> "mapped B")
                .addMapping(
                    input -> input.equals("input X"),
                    input -> "mapped X")
                .addMapping(
                    input -> input.equals("input C"),
                    input -> "mapped C")
                .toMapper();

        String expected = "mapped X";
        Optional<Long> actual = mapper.map("input X",Long.class );

        assertFalse(actual.isPresent());

    }

}

//////////////////////////////////////////////////////////////////////////////
